Code/GraphMol/ForceFieldHelpers/CrystalFF/TorsionPreferences.cpp

```diff
@@ -0,0 +1,674 @@
+          } // end loop over matches
+
+        } // end loop over patterns
+      }
+
+      // apply basic knowledge such as flat aromatic rings, other sp2-centers,
+      // straight triple bonds, etc.
+      if (useBasicKnowledge) {
+
+        BIT_SET doneAtoms(na);
+        ROMol::ADJ_ITER nbrIdx;
+        ROMol::ADJ_ITER endNbrs;
+
+        // inversion terms (improper torsions / out-of-plane bends / inversion)
+        // loop over atoms
+        for (int aid2 = 0; aid2 < na; ++aid2) {
+          if (!(doneAtoms[aid2])) {
+            std::vector<int> atoms(4, -1);
+            atoms[1] = aid2;
+            const Atom *atom2 = mol.getAtomWithIdx(atoms[1]);
+            int at2AtomicNum = atom2->getAtomicNum();
+
+            // if atom is a N,O or C and SP2-hybridized
+            if (((at2AtomicNum == 6) || (at2AtomicNum == 7) || (at2AtomicNum == 8))
+                && (atom2->getHybridization() == Atom::SP2)) {
+              // get neighbors
+              boost::tie(nbrIdx, endNbrs) = mol.getAtomNeighbors(atom2);
+              // check if enough neighbours
+              if (std::distance(nbrIdx,endNbrs) != 3) {
+                continue;
+              }
+              unsigned int i = 0;
+              unsigned int isBoundToSP2O = 0; // false
+              for (; nbrIdx != endNbrs; ++nbrIdx) {
+                const Atom *atomX = mol[*nbrIdx].get();
+                atoms[i] = atomX->getIdx();
+                // if the central atom is sp2 carbon and is bound to sp2 oxygen, set a flag
+                if (!isBoundToSP2O) {
+                  isBoundToSP2O = ((at2AtomicNum == 6) && (atomX->getAtomicNum() == 8)
+                    && (atomX->getHybridization() == Atom::SP2));
+                }
+                if (!i) {
+                  ++i;
+                }
+                ++i;
+              }
+              atoms.push_back(at2AtomicNum);
+              atoms.push_back(isBoundToSP2O);
+              improperAtoms.push_back(atoms);
+              /*if (verbose) {
+                std::cout << "out-of-plane bend: " << atoms[0] << " " << atoms[1] << " "
+                    << atoms[2] << " " << atoms[3] << std::endl;
+              }*/
+            }
+          } // if atom is a N,O or C and SP2-hybridized
+        }
+
+        // torsions for flat rings
+        const RingInfo *rinfo = mol.getRingInfo(); // FIX: make sure we have ring info
+        CHECK_INVARIANT(rinfo, "");
+        const VECT_INT_VECT &atomRings = rinfo->atomRings();
+        for (VECT_INT_VECT_CI rii = atomRings.begin(); rii != atomRings.end(); rii++) {
+          unsigned int rSize = rii->size();
+          // we don't need to deal with 3 membered rings
+          // and we do not treat rings greater than 6
+          if (rSize < 4 || rSize > 6) {
+            continue;
+          }
+          // loop over ring atoms
+          for (unsigned int i = 0; i < rSize; ++i) {
+            // proper torsions
+            aid1 = (*rii)[i];
```

Comment:
I stared at this block of code for a good 2-3 minutes thinking "WTF?" and then I figured it out. Nice one! :+1: